alf/utils/data_buffer.py

```diff
@@ -81,21 +72,18 @@ def add_batch(self, batch):
         Args:
             batch (Tensor): shape should be [batch_size] + tensor_space.shape
         """
-
-        with tf.device(self._device):
-            batch_size = get_nest_batch_size(batch, tf.int32)
-            n = tf.minimum(batch_size, self._capacity)
-            indices = tf.range(self._current_pos,
-                               self._current_pos + n) % self._capacity
-            indices = tf.expand_dims(indices, axis=-1)
-
-            tf.nest.map_structure(
-                lambda buf, bat: buf.scatter_nd_update(
-                    indices, tf.stop_gradient(bat[-n:])), self._buffer, batch)
-
-            self._current_pos.assign((self._current_pos + n) % self._capacity)
-            self._current_size.assign(
-                tf.minimum(self._current_size + n, self._capacity))
+        batch_size = alf.nest.get_nest_batch_size(batch)
+        with alf.device(self._device):
+            n = torch.min(self._capacity, torch.as_tensor(batch_size))
+            indices = torch.arange(self._current_pos,
+                                   self._current_pos + n) % self._capacity
+            alf.nest.map_structure(
+                lambda buf, bat: buf.__setitem__(indices, bat[-n:].detach()),
+                self._buffer, batch)
+
+            self._current_pos = (self._current_pos + n) % self._capacity
```

Comment:
use `self._current_pos.copy_()` so that the object `self._current_pos` refers to does not change,  which can reduce the overhead of book keeping.